apphub/foundation_model/lora/lora_torch.py

```diff
@@ -0,0 +1,570 @@
+                              num_heads=12,
+                              patch_size=16,
+                              qkv_bias=True,
+                              use_rel_pos=True,
+                              global_attn_indexes=(2, 5, 8, 11),
+                              window_size=14,
+                              out_chans=256)
+    if weights_path:
+        print("Loading SAM Encoder weights from {}...".format(weights_path))
+        load_model(encoder, weights_path=weights_path)
+    return encoder
+
+
+class LoRA_SamEncoder(nn.Module):
+    def __init__(self, r=4, base_encoder_weights=None, lora_weights=None):
+        super().__init__()
+        self.base_encoder = sam_encoder_b(weights_path=base_encoder_weights)
+        # create for storage, then we can init them or load weights
+        self.w_As = []  # These are linear layers
+        self.w_Bs = []
+        print('parameter number of image encoder: ' +
+              str(sum(p.numel() for p in self.base_encoder.parameters() if p.requires_grad)))
+        # lets freeze all image encoders
+        for param in self.base_encoder.parameters():
+            param.requires_grad = False
+        lora_param_num = 0
+        # Here, we do the surgery by replacing the qkv function of original attention
+        for blk in self.base_encoder.blocks:
+            w_qkv_linear = blk.attn.qkv
+            dim = w_qkv_linear.in_features
+            w_a_linear_q = nn.Linear(dim, r, bias=False)
+            w_b_linear_q = nn.Linear(r, dim, bias=False)
+            w_a_linear_v = nn.Linear(dim, r, bias=False)
+            w_b_linear_v = nn.Linear(r, dim, bias=False)
+            self.w_As.append(w_a_linear_q)
+            self.w_Bs.append(w_b_linear_q)
+            self.w_As.append(w_a_linear_v)
+            self.w_Bs.append(w_b_linear_v)
+            blk.attn.qkv = _LoRA_qkv(
+                w_qkv_linear,
+                w_a_linear_q,
+                w_b_linear_q,
+                w_a_linear_v,
+                w_b_linear_v,
+            )
+            lora_param_num += sum(p.numel() for p in blk.attn.qkv.parameters() if p.requires_grad)
+        print('parameter number of LoRA: ' + str(lora_param_num))
+        self.lora_modules = nn.ModuleList(self.w_As + self.w_Bs)
+        if lora_weights:
+            self.load_lora_weights(lora_weights)
+        else:
+            self.init_lora_weights()
+
+    def init_lora_weights(self):
+        for w_A in self.w_As:
+            nn.init.kaiming_uniform_(w_A.weight, a=math.sqrt(5))
+        for w_B in self.w_Bs:
+            nn.init.zeros_(w_B.weight)
+
+    def forward(self, x):
+        out = self.base_encoder(x)
+        return out
+
+    def load_lora_weights(self, save_path):
+        print("Loading LoRA weights from {}...".format(save_path))
+        self.lora_modules.load_state_dict(
+            torch.load(save_path, map_location='cpu' if torch.cuda.device_count() == 0 else None))
```

Comment:
use our util gpu detection method instead of device_count() for compatibility with Mac